packages/vite/src/node/plugins/define.ts

```diff
@@ -122,23 +119,34 @@ export function definePlugin(config: ResolvedConfig): Plugin {
         return
       }
 
-      const [define, pattern, banner] = ssr ? ssrPattern : defaultPattern
+      const [define, pattern, importMetaEnvVal] = ssr
+        ? ssrPattern
+        : defaultPattern
       if (!pattern) return
 
       // Check if our code needs any replacements before running esbuild
       pattern.lastIndex = 0
       if (!pattern.test(code)) return
 
+      let marker = importMetaEnvMarker
+      while (new RegExp(escapeRegex(marker)).test(code)) {
+        marker += '$'
+      }
+      if (marker !== importMetaEnvMarker && 'import.meta.env' in define) {
+        define['import.meta.env'] = marker
+      }
```

Comment:
I don't think we can mutate the `define` here like this as it will affect all other define. We'd have to do something like `define = { ...define, 'import.meta.env': marker }` here. It's not very performant, but it's unlikely that this would happen to many modules.